src/Stratis.Bitcoin/BlockPulling2/BlockPullerBehavior.cs

```diff
@@ -0,0 +1,245 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Interfaces;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.P2P.Protocol;
+using Stratis.Bitcoin.P2P.Protocol.Behaviors;
+using Stratis.Bitcoin.P2P.Protocol.Payloads;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.BlockPulling2
+{
+    /// <summary>
+    /// Relation of the node's puller to a network peer node.
+    /// Keeps all peer-related values that <see cref="BlockPuller"/> needs to know about a peer.
+    /// </summary>
+    public interface IBlockPullerBehavior : INetworkPeerBehavior
+    {
+        /// <summary>Relative quality score of a peer.</summary>
+        /// <remarks>It's a value from <see cref="MinQualityScore"/> to <see cref="MaxQualityScore"/>.</remarks>
+        double QualityScore { get; }
+
+        /// <summary>Upload speed of a peer in bytes per second.</summary>
+        int SpeedBytesPerSecond { get; }
+
+        /// <summary>Tip claimed by peer.</summary>
+        ChainedHeader Tip { get; set; }
+
+        /// <summary>
+        /// Adds peer performance sample that is used to estimate peer's qualities.
+        /// </summary>
+        /// <param name="blockSizeBytes">Block size in bytes.</param>
+        /// <param name="delaySeconds">Time in seconds it took peer to deliver a block.</param>
+        void AddSample(long blockSizeBytes, double delaySeconds);
+
+        /// <summary>Applies a penalty to a peer for not delivering a block.</summary>
+        /// <param name="delaySeconds">Time in which peer didn't deliver assigned blocks.</param>
+        /// <param name="notDeliveredBlocksCount">Number of blocks peer failed to deliver.</param>
+        void Penalize(double delaySeconds, int notDeliveredBlocksCount);
+
+        /// <summary>Called when IBD state changed.</summary>
+        void OnIbdStateChanged(bool isIbd);
+
+        /// <summary>Recalculates the quality score for this peer.</summary>
+        /// <param name="bestSpeedBytesPerSecond">Speed in bytes per second that is considered to be the maximum speed.</param>
+        void RecalculateQualityScore(int bestSpeedBytesPerSecond);
+
+        /// <summary>Requests blocks from this peer.</summary>
+        /// <param name="hashes">Hashes of blocks that should be asked to be delivered.</param>
+        /// <exception cref="OperationCanceledException">Thrown in case peer is in the wrong state or TCP connection was closed during sending a message.</exception>
+        Task RequestBlocksAsync(List<uint256> hashes);
+    }
+
+    /// <inheritdoc cref="IBlockPullerBehavior"/>
+    public class BlockPullerBehavior : NetworkPeerBehavior, IBlockPullerBehavior
+    {
+        internal const double MinQualityScore = 0.01;
```

Comment:
I guess the reason for decimals is the granularity if the score?

Reply:
Yes, correct, it's to be more precise. 

Now we have quality score from 0 to 1 where 1 means the best speed of all peers.